### PR TITLE
Fix usage of constant in xml namespace uri definitions

### LINKED ISSUE #35277: `xmlns xml:XMLNS_NAMESPACE_URI as ns0` seem to not work
**Description:**
$subject.

```bal
xmlns xml:XMLNS_NAMESPACE_URI as ns0;

public function main() {
    string s = ns0:foo;
    io:println(s); // prints "{}foo"
}
```

### PR DESCRIPTION
**Pull request by poorna2152**
## Purpose
Fixes #35277

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 1]
```diff
@@ -1166,14 +1166,13 @@ public void initPredeclaredModules(Map<Name, BPackageSymbol> predeclaredModules,
...
+        String nsURI = "";
         if (xmlnsNode.namespaceURI.getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
             BLangSimpleVarRef varRef = (BLangSimpleVarRef) xmlnsNode.namespaceURI;
-            if (missingNodesHelper.isMissingNode(varRef.variableName.value)) {
-                nsURI = "";
-            } else {
-                // TODO: handle const-ref (#24911)
-                nsURI = "";
+            if (!missingNodesHelper.isMissingNode(varRef.variableName.value)) {
+                if (Symbols.isFlagOn(varRef.symbol.flags, Flags.CONSTANT)) {
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1173]: We can combine these two conditions

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 2]
```diff
@@ -1166,14 +1166,13 @@ public void initPredeclaredModules(Map<Name, BPackageSymbol> predeclaredModules,
...
+                    nsURI = String.valueOf(((BConstantSymbol) varRef.symbol).value);
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1174]:
```suggestion
                    nsURI = ((BConstantSymbol) varRef.symbol).value.toString();
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xmlns-def.bal [hunk 3]
```diff
@@ -0,0 +1,20 @@
+const A = "http://ballerina.com/a";
+
+xmlns A as ns0;
```
Comment (prakanth97) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xmlns-def.bal:3]: Shall we add a test where const A is union type(one member is string). Shall we add a negative test where A is not a string type?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 4]
```diff
@@ -1166,14 +1166,12 @@ public void initPredeclaredModules(Map<Name, BPackageSymbol> predeclaredModules,
...
-        String nsURI;
+        String nsURI = "";
         if (xmlnsNode.namespaceURI.getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
             BLangSimpleVarRef varRef = (BLangSimpleVarRef) xmlnsNode.namespaceURI;
-            if (missingNodesHelper.isMissingNode(varRef.variableName.value)) {
-                nsURI = "";
-            } else {
-                // TODO: handle const-ref (#24911)
-                nsURI = "";
+            if (!missingNodesHelper.isMissingNode(varRef.variableName.value) &&
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1172]: Why do we require this check `!missingNodesHelper.isMissingNode(varRef.variableName.value)`? I believe we may no longer need this check.

Comment (poorna2152) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Fixed

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/xml/XMLNSDefTest.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * Test class for XMLNS Definitions.
+ *
+ * @since 2201.9.0
+ */
+public class XMLNSDefTest {
+
+    private CompileResult result;
+    private CompileResult negativeResult;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/xml/XMLNSDefTest.java:37]: Can be local variables in the test methods since they are used only once.

File: tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xmlns-def-negative.bal [hunk 6]
```diff
@@ -0,0 +1,6 @@
+const A = 1;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xmlns-def-negative.bal:1]: Add licence headers.

File: tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xmlns-def.bal [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+    string s4 = ns3:bar;
+    assert(s4, "{http://ballerina.com/c}bar");
+}
+
+function assert(anydata actual, anydata expected) {
+    if (expected != actual) {
+        string reason = "expected `" + expected.toString() + "`, but found `" + actual.toString() + "`";
+        error e = error(reason);
+        panic e;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xmlns-def.bal:29]: I'd merge these into a single line.

File: tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xmlns-def.bal [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    string s2 = ns1:foo;
+    assert(s2, "{http://www.w3.org/2000/xmlns/}foo");
+
+    string s3 = ns2:bar;
+    assert(s3, "{http://ballerina.com/b}bar");
+
+    string s4 = ns3:bar;
+    assert(s4, "{http://ballerina.com/c}bar");
+}
+
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xmlns-def.bal:24]: Let's add tests with module-level variables too.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 4 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: This check can still evaluate to true and generally help with accuracy, although probably not specifically with the const symbol check.

```ballerina
xmlns as ns0;
```

Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: If a node is missing, we are issuing an error in the parser. What is the reason for rechecking this on the compiler side?

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Compilation will still continue when possible. As it does with the snippet I've shared.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 9]
```diff
@@ -1166,14 +1166,11 @@ public void initPredeclaredModules(Map<Name, BPackageSymbol> predeclaredModules,
...
-            if (missingNodesHelper.isMissingNode(varRef.variableName.value)) {
-                nsURI = "";
-            } else {
-                // TODO: handle const-ref (#24911)
-                nsURI = "";
+            if (Symbols.isFlagOn(varRef.symbol.flags, Flags.CONSTANT)) {
+                nsURI = ((BConstantSymbol) varRef.symbol).value.toString();
             }
         } else {
             nsURI = (String) ((BLangLiteral) xmlnsNode.namespaceURI).value;
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1176]: With this, an error is given for `xmlns "" as ex1;` with the message `cannot bind prefix ex1 to the empty namespace name`. Please verify this behavior is correct or not?. If so, we should give an error for the following scenario as well
```ballerina
const string variable = "";
xmlns variable as ex;
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xmlns-def.bal (hunk 8 above)
Comment (poorna2152) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xmlns-def.bal:24]: In that case would it make sense to move this test file into the `bala/types` tests or create a new file for xmlns definitions in the bala/types directory.

File: tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xmlns-def.bal
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xmlns-def.bal:1]: What you have atm is xmlns-decl.

We needs tests with xmlns-decl-stmt too.

Let's rename the test/file as xmlns tests.

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xmlns-def.bal:1]: Also for bal file names we use underscores.

File: tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xmlns-def.bal (hunk 8 above)
Comment (poorna2152) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xmlns-def.bal:24]: Sorry, misunderstood the module-level variables with variables inside modules.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 4 above)
Comment (poorna2152) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: We can skip this condition here right. Since we are by default setting the value here to "".

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 9 above)
Comment (poorna2152) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1177]: Updated

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 10]
```diff
@@ -1210,6 +1206,12 @@ public void visit(BLangXMLNS xmlnsNode) {
         defineSymbol(xmlnsNode.prefix.pos, xmlnsSymbol);
     }
 
+    private void checkInvalidNameSpaceDeclaration(Location pos, BLangIdentifier prefix, String nsURI) {
+        if (!nullOrEmpty(prefix.value) && nsURI.isEmpty()) {
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1210]: Why do we need `nullOrEmpty(prefix.value)` check?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 11]
```diff
@@ -1166,20 +1166,16 @@ public void initPredeclaredModules(Map<Name, BPackageSymbol> predeclaredModules,
...
+            if (Symbols.isFlagOn(varRef.symbol.flags, Flags.CONSTANT)) {
+                nsURI = ((BConstantSymbol) varRef.symbol).value.toString();
+                checkInvalidNameSpaceDeclaration(xmlnsNode.pos, xmlnsNode.prefix, nsURI);
             }
         } else {
             nsURI = (String) ((BLangLiteral) xmlnsNode.namespaceURI).value;
-            if (!nullOrEmpty(xmlnsNode.prefix.value) && nsURI.isEmpty()) {
-                dlog.error(xmlnsNode.pos, DiagnosticErrorCode.INVALID_NAMESPACE_DECLARATION, xmlnsNode.prefix);
-            }
+            checkInvalidNameSpaceDeclaration(xmlnsNode.pos, xmlnsNode.prefix, nsURI);
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1178]: Shall we do this check after the if statement? That way, we can perform the `checkInvalidNameSpaceDeclaration` check in a single location, right after assigning a value for `nsURI`.

**Comment by chiranSachintha** [comment]
Better if we can add a bala test for negative scenario as well.

module/Foo 
```ballerina
public const k = "";
```
module/Bar

```ballerina
xmlns Foo:k as ns0;

public function main() {
    string s = ns0:foo;
}
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 10 above)
Comment (poorna2152) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1210]: This check is used to allow the following scenario,
```ballerina
xmlns "";
```
When xmlns is defined without a prefix and has an empty string for the namespace-uri. Based on the [xml spec](https://www.w3.org/TR/xml-names/#iri-use),
this seems to be an issue as,
```
The empty string, though it is a legal URI reference, cannot be used as a namespace name.
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 11 above)
Comment (poorna2152) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1178]: This is needed in two places because we don't handle the case where the namespace uri is defined using a variable reference and which is not a constant reference. In that case also we set the namespaceURI to be empty string. We decided to handle that case in a separately (different PR). So we need to have two checks here. Otherwise an invalid error would be thrown.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 10 above)
Comment (poorna2152) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1210]: Created [issue](https://github.com/ballerina-platform/ballerina-lang/pull/41958) to track this

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 12]
```diff
@@ -1166,20 +1166,16 @@ public void initPredeclaredModules(Map<Name, BPackageSymbol> predeclaredModules,
 
     @Override
     public void visit(BLangXMLNS xmlnsNode) {
-        String nsURI;
+        String nsURI = "";
```
Comment (SasinduDilshara) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1169]: Do we need this change, we can initialise this as null.

Comment (poorna2152) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1169]: This was by default initialized to `""` due to this [bug](https://github.com/ballerina-platform/ballerina-lang/issues/42075) and because we decided to handle this separately.

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/types/xmlns_bala_test.bal [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+import testorg/foo;
+
+xmlns foo:XMLNSA as ns0;
+xmlns foo:XMLNSB as ns1;
+
+function testXMLNSDeclUsingModuleConstant() {
+    string s1 = ns0:foo;
```
Comment (chiranSachintha) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/types/xmlns_bala_test.bal:23]: Is there a specific reason for doing this assignment? If not, we can directly use `ns0:foo` for assertion without assigning it.

Comment (poorna2152) [reply tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/types/xmlns_bala_test.bal:23]: Fixed

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/XMLNSBalaTest.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import static org.ballerinalang.test.BAssertUtil.validateError;
+
+/**
+ * Bala test cases for xmlns declaration.
+ *
+ * @since 2201.9.0
```
[the remark this answers is not in the document]
Comment (poorna2152) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/XMLNSBalaTest.java:34]: It's optional and it is better to have in this case IMO because this fix is intended for the master.

**Review by chiranSachintha — DISMISSED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 15]
```diff
@@ -1166,20 +1166,16 @@ public void initPredeclaredModules(Map<Name, BPackageSymbol> predeclaredModules,
...
         if (xmlnsNode.namespaceURI.getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
             BLangSimpleVarRef varRef = (BLangSimpleVarRef) xmlnsNode.namespaceURI;
-            if (missingNodesHelper.isMissingNode(varRef.variableName.value)) {
-                nsURI = "";
-            } else {
-                // TODO: handle const-ref (#24911)
-                nsURI = "";
+            if (Symbols.isFlagOn(varRef.symbol.flags, Flags.CONSTANT)) {
+                nsURI = ((BConstantSymbol) varRef.symbol).value.toString();
+                checkInvalidNameSpaceDeclaration(xmlnsNode.pos, xmlnsNode.prefix, nsURI);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java:1174]: Can't we move this out the if/else blocks?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/XMLNSBalaTest.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    @Test
+    public void testXMLNSDeclUsingModuleConstant() {
+        BRunUtil.invoke(result, "testXMLNSDeclUsingModuleConstant");
+    }
+
+    @Test
+    public void testXMLNSDeclStmtUsingModuleConstant() {
+        BRunUtil.invoke(result, "testXMLNSDeclStmtUsingModuleConstant");
+    }
+
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/XMLNSBalaTest.java:55]: Shall we use a data provider instead?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/XMLNSBalaTest.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+    @Test
+    public void testXMLNSNegativeDefinition() {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/XMLNSBalaTest.java:57]:
```suggestion
    public void testXMLNSDefinitionNegative() {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/types/xmlns_bala_test.bal [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+    assert(ns1:foo, "{http://ballerina.com/b}foo");
+}
+
+function testXMLNSDeclStmtUsingModuleConstant() {
+    xmlns foo:XMLNSC as ns2;
+    assert(ns2:baz, "{http://ballerina.com/c}baz");
+}
+
+function assert(anydata actual, anydata expected) {
+    if (expected != actual) {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/types/xmlns_bala_test.bal:33]:
```suggestion
    if expected != actual {
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/XMLNSBalaTest.java
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/XMLNSBalaTest.java:1]: types isn't the correct package for this, is it? Shall we move to appropriate package? E.g., xmlns or declarations?

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/XMLNSBalaTest.java:1]: Check for other files too.

Comment (poorna2152) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/XMLNSBalaTest.java:1]: Existing folders under `test/bala/` [[1]](https://github.com/ballerina-platform/ballerina-lang/tree/master/tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala) doesn't have a folder that we can use to move this test to IMO. WDYT about creating a new folder `test/bala/declarations` and moving the `XMLNSBalaTest` there and also creating directory `test/declarations` and moving the `XMLNSTest` to.

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/bala/types/XMLNSBalaTest.java:1]: Yeah, we can create new ones.

**Review by MaryamZi — APPROVED** [review]

**Review by chiranSachintha — APPROVED** [review]
